### PR TITLE
Feature/implicit group

### PR DESCRIPTION
**Pull request by souziel**

### CODE REVIEW

File: server/src/forwarders/socketForwarder.ts [hunk 1]
```diff
@@ -177,6 +186,21 @@ class SocketForwarder {
...
+                this.observers.forEach((ob) => {
+                    if (ob.onAreaRemoved) {
+                        ob.onAreaRemoved(data.event?.data?.area_id);
+                    }
+                });
+            } else if (data?.event?.data?.action === 'update') {
+                this.observers.forEach((ob) => {
+                    if (ob.onAreaUpdated) {
+                        ob.onAreaUpdated(data.event?.data?.area_id);
+                    }
```
Comment (guyts96) [inline server/src/forwarders/socketForwarder.ts]: Why don't you use the notifyObservers method here ?

File: server/src/services/groupService.ts [hunk 2]
```diff
@@ -103,36 +129,89 @@ class GroupService implements EventObserver {
...
+
+    async generateImplicitGroup() {
+        // Generate implicit group per room
+        const rooms = await roomService.getRooms();
+        await Promise.all(rooms.map((r: Room) => this.generateImplicitGroupForOneRoom(r)));
+        // Generate implicit group for from all devices
+        const devices: Device[] = await deviceService.getDevices();
+        const entities = devices.flatMap((d: Device) => d.entities.filter((entity: Entity) => entity.id.startsWith('switch') || entity.id.startsWith('light')).map((entity) => entity.id));
+        if (!entities) {
+            return Promise.resolve();
```
Comment (guyts96) [inline server/src/services/groupService.ts]: Why resolve the promise when there are no entities ? Shouldn't we reject it ?

File: server/src/services/roomService.ts [hunk 3]
```diff
@@ -42,13 +42,18 @@ class RoomService {
...
     }
 
-    async getRoomById(areaId: string) {
+    async getRoomById(roomId: string): Promise<Room> {
         const rooms: Room[] = await this.getRooms();
-        return rooms.find((r: Room) => r.roomId === areaId);
+        console.log(roomId);
+        const room = rooms.find((r: Room) => r.roomId === roomId);
+        if (!room) {
+            throw new Error('There are no rooms with this id ');
```
Comment (guyts96) [inline server/src/services/roomService.ts]: Can we add the ID in the error message ?

File: server/src/types/observer.ts [hunk 4]
```diff
@@ -7,6 +7,8 @@ export interface EventObserver {
     onAutomationUpdated?: (data: HaEntityUpdated) => void;
     onEntityRegistryUpdated?: () => void;
     onDeviceRegistryUpdated?: () => void;
+    onAreaUpdated?: (roomId: string) => void;
+    onAreaRemoved?: (roomId: string) => void;
 }
 
 export type Event = 'authOk' | 'entityUpdated' | 'automationUpdated' | 'entityRegistryUpdated' | 'deviceRegistryUpdated';
```
Comment (guyts96) [inline server/src/types/observer.ts]: Shouldn't we add new events for onAreaUpdated and onAreaRemoved ?

**Review by guyts96 — APPROVED** [review]

**Review by asoran — APPROVED** [review]
🍇

**Pull request merged**